crates/utils/src/interval_map/mod.rs

```diff
@@ -0,0 +1,804 @@
+    }
+
+    fn set_parent(&self, node: NodeRef<T, V>) {
+        let _ignore = self.borrow_mut().parent.replace(node);
+    }
+}
+
+#[allow(clippy::missing_docs_in_private_items)]
+impl<T, V> NodeRef<T, V> {
+    fn max<F, R>(&self, op: F) -> R
+    where
+        F: FnOnce(&T) -> R,
+    {
+        op(self.borrow().max())
+    }
+
+    fn interval<F, R>(&self, op: F) -> R
+    where
+        F: FnOnce(&Interval<T>) -> R,
+    {
+        op(self.borrow().interval())
+    }
+
+    fn left<F, R>(&self, op: F) -> R
+    where
+        F: FnOnce(&NodeRef<T, V>) -> R,
+    {
+        op(self.borrow().left())
+    }
+
+    fn right<F, R>(&self, op: F) -> R
+    where
+        F: FnOnce(&NodeRef<T, V>) -> R,
+    {
+        op(self.borrow().right())
+    }
+
+    fn parent<F, R>(&self, op: F) -> R
+    where
+        F: FnOnce(&NodeRef<T, V>) -> R,
+    {
+        op(self.borrow().parent())
+    }
+
+    fn grand_parent<F, R>(&self, op: F) -> R
+    where
+        F: FnOnce(&NodeRef<T, V>) -> R,
+    {
+        let mut grand_parent = self.borrow().parent().borrow().parent_owned();
+        op(&mut grand_parent)
+    }
+}
+
+/// The color of the node
+#[derive(Debug, Clone, Copy)]
+enum Color {
+    /// Red node
+    Red,
+    /// Black node
+    Black,
+}
+
+/// The Interval stored in `IntervalMap`
+/// Represents the interval [low, high)
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Interval<T> {
+    /// Low value
+    low: T,
+    /// high value
+    high: T,
+}
+
+impl<T: Ord> Interval<T> {
+    /// Creates a new `Interval`
+    ///
+    /// # Panics
+    ///
+    /// This method panics when low is greater than high
+    #[inline]
+    pub fn new(low: T, high: T) -> Self {
+        assert!(low < high, "invalid range");
+        Self { low, high }
+    }
+
+    /// Checks if self overlaps with other interval
+    fn overlap(&self, other: &Self) -> bool {
+        self.high > other.low && other.high > self.low
+    }
+}
+
+/// A view into a single entry in a map, which may either be vacant or occupied.
+#[allow(missing_debug_implementations, clippy::exhaustive_enums)]
+pub enum Entry<'a, T, V> {
+    /// An occupied entry.
+    Occupied(OccupiedEntry<T, V>),
+    /// A vacant entry.
+    Vacant(VacantEntry<'a, T, V>),
+}
+
+/// A view into an occupied entry in a `IntervalMap`.
+/// It is part of the [`Entry`] enum.
+#[allow(missing_debug_implementations)]
+pub struct OccupiedEntry<T, V> {
+    /// The entry node
+    node: NodeRef<T, V>,
+}
+
+/// A view into a vacant entry in a `IntervalMap`.
+/// It is part of the [`Entry`] enum.
+#[allow(missing_debug_implementations)]
+pub struct VacantEntry<'a, T, V> {
+    /// Mutable reference to the map
+    map_ref: &'a mut IntervalMap<T, V>,
+    /// The interval of this entry
+    interval: Interval<T>,
+}
+
+impl<T, V> Entry<'_, T, V>
+where
+    T: Ord + Clone,
+{
+    /// Ensures a value is in the entry by inserting the default if empty, and returns
```

Comment:
Please modify the comment. This method doesn't return anything.

Reply:
now it returns the reference